### PR TITLE
Fix utilization URL for special characters

### PR DESCRIPTION
**Pull request by mayorova**
This is related to [THREESCALE-9033](https://issues.redhat.com/browse/THREESCALE-9033).

Previously https://github.com/3scale/pisoni/pull/31 fixed the same issue for managing application keys.

This PR fixes it for utilization and referrer filters endpoints.

### CODE REVIEW

File: lib/3scale/core/application_referrer_filter.rb [hunk 1]
```diff
@@ -25,7 +25,7 @@ def self.delete(service_id, application_id, value)
       end
 
       def self.base_uri(service_id, application_id)
-        "#{default_uri}#{service_id}/applications/#{application_id}/referrer_filters"
+        "#{default_uri}#{service_id}/applications/#{CGI::escape(application_id.to_s)}/referrer_filters"
```
Comment (akostadinov) [inline lib/3scale/core/application_referrer_filter.rb]: I wonder whether we should use `ERB::Util.url_encode` or `CGI::escape`.

```
[1] pry(main)> str = "asdasd asdas"
[3] pry(main)> require 'cgi'
[6] pry(main)> require 'erb'

[4] pry(main)> CGI.escape str
=> "asdasd+asdas"

[7] pry(main)> ERB::Util.url_encode str
=> "asdasd%20asdas"
```

I'm not sure what is the processing on the other side. So this is just a question. Not an objection.

Comment (akostadinov) [reply lib/3scale/core/application_referrer_filter.rb]: Rails `Hash#to_query` seems to be using `CGI.escape` internally, so probably it is fine. But good to test it anyway.

Comment (mayorova) [reply lib/3scale/core/application_referrer_filter.rb]: Yeah, I was looking for the best suited encoder when working on https://github.com/3scale/pisoni/pull/31, and I ended up with `CGI::escape`, but I don't remember the reason :sweat_smile: 
Probably one of the points was that it was already being used in other parts of the library (and worked successfully), so I decided not to innovate. Or maybe there was a different reason.

If you are really interested, I can dig into it (although I would prefer to avoid it :grimacing: )

Comment (akostadinov) [reply lib/3scale/core/application_referrer_filter.rb]: I think it has to be tested which one would work properly with a space character in the key. No need to dig further because this is the only difference that I was able to identify.

Comment (mayorova) [reply lib/3scale/core/application_referrer_filter.rb]: Well, yes, I had tested that `CGI::escape` works as expected (in a manual integration test).
However, I decided to investigate this further to clarify any doubts.

This `pisoni` client is used for apisonator API, so, I checked what apisonator does with encoded spaces.

1. its base, sinatra, unescapes URL-encoded values in captures like this: https://github.com/sinatra/sinatra/blob/v2.2.4/lib/sinatra/base.rb#L1069

```
[5] pry(main)> URI::Parser.new.unescape("space%20man")
=> "space man"
[6] pry(main)> URI::Parser.new.unescape("space+man")
=> "space+man"
```
As we can see this results in different values of the `:app_id` [capture](https://github.com/3scale/apisonator/blob/master/app/api/internal/utilization.rb#L4) in the utilization URL (that this PR fixes).

2. When the application is fetched from Redis, the key for Redis passes through [encoding](https://github.com/3scale/apisonator/blob/8e7ff4f28088a0b78c173c90ad267524e532b76a/lib/3scale/backend/storage_key_helpers.rb#L4-L6) that replaces all spaces with `+`.

This is why both `%20` and `+` have the same effect.

By the way, regarding whether or not `+` should be decoded to spaces by sinatra, there was actually some discussion about it. It was implemented at some point, but then [reverted](https://github.com/sinatra/sinatra/commit/b4d7d4bb1088f5a6caf1bfbbba80717db9232fe7). And finally, `URI.decode` was also changed to `URI.unescape` [here](https://github.com/sinatra/sinatra/commit/4a02a757fde29ba21d241c838c8114f64a393fae).

File: docker/docker-compose.apisonator.yml [hunk 2]
```diff
@@ -12,10 +12,10 @@ services:
     command:
       3scale_backend start -p 3000 -l "${APISONATOR_LOGFILE}"
     environment:
-      RACK_ENV: development
+      RACK_ENV: production
       APISONATOR_IAPI_USER: "${APISONATOR_IAPI_USER}"
       APISONATOR_IAPI_PASSWORD: "${APISONATOR_IAPI_PASSWORD}"
-      APISONATOR_REDIS_URL: redis://redis:6379
+      APISONATOR_REDIS_URL: redis://redis:6379git
```
Comment (akostadinov) [inline docker/docker-compose.apisonator.yml]: git ftw!

File: lib/3scale/core/application_referrer_filter.rb (hunk 1 above)
Comment (akostadinov) [reply lib/3scale/core/application_referrer_filter.rb]: To confirm, this means we should use `ERB::Util.url_encode` to match sinatra behavior, correct?

Because `CGI.escape str` encodes spaces to `+`. But then how manual tests let it still work? Maybe porta also used `CGI.escape str` when sending to backend and the double mistake was negated? :)

File: docker/docker-compose.apisonator.yml (hunk 2 above)
Comment (mayorova) [reply docker/docker-compose.apisonator.yml]: hahaa :sweat_smile:

File: lib/3scale/core/application_referrer_filter.rb (hunk 1 above)
Comment (mayorova) [reply lib/3scale/core/application_referrer_filter.rb]: @akostadinov Mmm... no :upside_down_face: 

> To confirm, this means we should use `ERB::Util.url_encode` to match sinatra behavior, correct?

How did you come to this conclusion based on the info above? :sweat_smile: 

So, to summarize, for the app ID `space man` here are the transformations:

| encoder                      | pisoni URL     | sinatra capture | redis key   |
|----|----|----|----|
| using `ERB::Util.url_encode` | `/space%20man` | `space man`     | `space+man` |
| using `CGI::escape`          | `/space+man`   | `space+man`     | `space+man` |

That's why it works - the outcome (the key to search for the application in Redis) is exactly the same. If anything, I'd prefer to have it with `+` directly, to avoid two transformations.

By the way, it is a hypothetical case, because we don't allow spaces for application IDs, see [here](https://github.com/3scale/porta/blob/404885a5a00937c1329c219bf03e1559c11b31af/app/models/cinstance.rb#L124-L125).

File: docker/docker-compose.apisonator.yml [hunk 3]
```diff
@@ -12,7 +12,7 @@ services:
     command:
       3scale_backend start -p 3000 -l "${APISONATOR_LOGFILE}"
     environment:
-      RACK_ENV: development
+      RACK_ENV: test
```
Comment (mayorova) [inline docker/docker-compose.apisonator.yml:15]: With `RACK_ENV=development` there was an issue with loading `dotenv` (because the gem not present in the image).

With `RACK_ENV=production` there is another issue, and it's that some endpoints are meant to be used for test and development only, so they are disabled in production, which makes the tests fail, specifically [this one](https://github.com/3scale/apisonator/blob/8e7ff4f28088a0b78c173c90ad267524e532b76a/app/api/internal/events.rb#L23-L37), which only works in [test and development](https://github.com/3scale/apisonator/blob/8e7ff4f28088a0b78c173c90ad267524e532b76a/app/api/internal/internal.rb#L44-L46).

```
ThreeScale::Core::Event::.delete::with an existing id event#test_0001_returns true when deleting an existing event:
ThreeScale::Core::APIClient::APIError: ThreeScale::Core::APIClient::APIError: 404 POST /internal/events/, attributes: {:status=>"not_found", :error=>"Not found"}, response.body: {"status":"not_found","error":"Not found"}
```

File: lib/3scale/core/application_referrer_filter.rb (hunk 1 above)
Comment (akostadinov) [reply lib/3scale/core/application_referrer_filter.rb]: Got it why it works.

The point in using `ERB::Util.url_encode` is that we use same standard to encode on client and decode on server.

What goes to redis from server is server's business. If server is modified to use a different encoding to database, that shouldn't affect how client communicates with server. e.g. server might be changed to write Base64 in database or whatever else.

Also I understand about spaces not allowed. But it is only for sanity's sake that we use same encoding on client as we do on server. Actually wait, ideally we use `URI::Parser.new.escape` instead of `ERB::Util.url_encode`. Although better avoid creating a new parser object every time :)

Comment (mayorova) [reply lib/3scale/core/application_referrer_filter.rb]: @akostadinov OK, you convinced me that spaces need to be encoded as `%20`. Well, I was convinced by reading some blog posts and comments, specifically this one https://stackoverflow.com/a/29948396/955810 (and its source).
I'll fix this for all places where special characters may appear in the URL path.

Comment (akostadinov) [reply lib/3scale/core/application_referrer_filter.rb]: So what is better `URI::Parser.new.escape` or `ERB::Util.url_encode`?

Comment (akostadinov) [reply lib/3scale/core/application_referrer_filter.rb]: I'm not sure the blog is entirely correct. It claims that `/` is alright in query but it is listed as reserved in both - [rfc1738](https://datatracker.ietf.org/doc/html/rfc1738#section-3.3) and [rfc2396](https://datatracker.ietf.org/doc/html/rfc2396#section-3.4)

The common characters to be escaped are also from [rfc2396](https://datatracker.ietf.org/doc/html/rfc2396#section-2.4.3)

Anyway, it appears that `URI::Parser` is https://ruby-doc.org/stdlib-3.1.0/libdoc/uri/rdoc/URI/RFC2396_Parser.html

But still it doesn't escape the `/` character. Which is strange, or I misinterpret the standard somehow...
```
[8] pry(main)> URI::RFC2396_Parser.new.escape "assad/ads da"
=> "assad/ads%20da"
```

Comment (akostadinov) [reply lib/3scale/core/application_referrer_filter.rb]: While `ERB::Util.url_encode` uses under the hood [CGI.escapeURIComponent(string)](https://rubyapi.org/3.2/o/cgi/escape#method-i-escapeURIComponent) which is supposed to follow [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.4) which in turn supersedes [rfc2396](https://datatracker.ietf.org/doc/html/rfc2396#section-3.4) and clarifies it.

```
[6] pry(main)> ERB::Util.url_encode "assad/ads da"
=> "assad%2Fads%20da"
```

But funnily from that same spec:
>  The characters slash ("/") and question mark ("?") may represent data
>  within the query component.  Beware that some older, erroneous
>  implementations may not handle such data correctly when it is used as
>  the base URI for relative references (Section 5.1), apparently
>  because they fail to distinguish query data from path data when
>  looking for hierarchical separators.  However, as query components
>  are often used to carry identifying information in the form of
>  "key=value" pairs and one frequently used value is a reference to
>  another URI, it is sometimes better for usability to avoid percent-
>  encoding those characters.

Basically the rfc2396 parser not encoding them while rfc3986 parser encoding them against it's own recommendation.

I think the main issue is that there are no separate methods to encode/decode query, path and whole URL.

So it is a little unclear which parts of the URLs the different methods are supposed to handle.

In this case either option appears to be alright.

Comment (mayorova) [reply lib/3scale/core/application_referrer_filter.rb]: This is all very confusing indeed :zany_face:

Comment (mayorova) [reply lib/3scale/core/application_referrer_filter.rb]: Hm, so I'm really lost in all those RFC, which characters are allowed in which part of the URL, and what each library implements correctly, and what they implement wrong.

What I did notice is that `URI::Parser.new.escape` doesn't work well for some characters, which luckily we had in some test scenarios.

Compare:

```
[7] pry(main)> str = 'a bc{1}$3?'
=> "a bc{1}$3?"
[8] pry(main)> URI::Parser.new.escape str
=> "a%20bc%7B1%7D$3?"
[9] pry(main)> ERB::Util.url_encode str
=> "a%20bc%7B1%7D%243%3F"
[10] pry(main)> CGI::escape str
=> "a+bc%7B1%7D%243%3F"
```

`URI::Parser` doesn't encode `$` - which is probably fine and `?` - which is actually a problem, because it makes the server interpret this `?` as the end of path segment, and the beginning of query string.

So, I think `ERB::Util.url_encode` might actually be the most suited for this job.

Comment (akostadinov) [reply lib/3scale/core/application_referrer_filter.rb]: Ok, agreed to your conclusion. 

In fact [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.4) specifically allows `?` in query part as I posted above and suggests that in the past it was only erroneously parsed wrong. But apparently Ruby implementation of [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.4) accounts for erroneous parsing in ruby world :)

Thank you for looking into this!

Comment (jlledom) [reply lib/3scale/core/application_referrer_filter.rb]: What we want is to percent-encode according to RFC3986, which seems to be the current standard. But it seems ruby doesn't really support it: https://bugs.ruby-lang.org/issues/18822

Probably `ERB::Util.url_encode` is the best option

Comment (jlledom) [reply lib/3scale/core/application_referrer_filter.rb]: @mayorova shouldn't we update also the changes from #31 ?

Comment (mayorova) [reply lib/3scale/core/application_referrer_filter.rb]: @jlledom yes, I'll do a mass replacement of all encoding, once I confirm that `ERB::Util.url_encode` is really the way to go :sweat_smile: right now having issues with some other special characters, e.g. `\`

Comment (akostadinov) [reply lib/3scale/core/application_referrer_filter.rb]: last comment of https://bugs.ruby-lang.org/issues/18822 suggests to use 

> But it's not useless since RFC 2396 was a Draft Standard and was obsoleted and updated by RFC 3986 which is an Internet Standard as CGI.escapeURIComponent and CGI.unescapeURIComponent implements RFC 398

And what I see is that `ERB::Util.url_encode` uses it under the hood. We may use that directly. The only "issue" I see with that is that `/` and `?` are % encoded against the recommendation but not violating the standard.

Comment (mayorova) [reply lib/3scale/core/application_referrer_filter.rb]: > And what I see is that `ERB::Util.url_encode` uses it under the hood. We may use that directly. The only "issue" I see with that is that `/` and `?` are % encoded against the recommendation but not violating the standard.

What recommendation are you referring to?

I think for path segments it makes total sense to %-encode `/` and `?`, because they have a special meaning and otherwise modify the path (terminating the current path segment).

Comment (akostadinov) [reply lib/3scale/core/application_referrer_filter.rb]: Ah, this is path segment, not query. Ok.

Comment (mayorova) [reply lib/3scale/core/application_referrer_filter.rb]: @akostadinov @jlledom fixed in https://github.com/3scale/pisoni/pull/33/commits/59751c8194a998aaf3c1985a357cd23a0b47fae6

Now, I am not sure about this path traversal thing, 
https://sinatrarb.com/protection/
https://github.com/sinatra/sinatra/blob/v2.2.4/rack-protection/lib/rack/protection/path_traversal.rb#L28-L29

This doesn't not allow us to use `/` and `\` in IDs/keys. It decodes them before routing takes place (and replaces `\` with `/`) so the requests are not routed properly and result in 404.

However, I am not quite sure what potential security implications may disabling it have, e.g.
```
use Rack::Protection, :except => :path_traversal
```
Ideas?

Comment (akostadinov) [reply lib/3scale/core/application_referrer_filter.rb]: Could you elaborate what the issue is? Encoding of the element should ensure no unescaped slashes are present so we will get the value with the slashes after rack URL processing.

Comment (mayorova) [reply lib/3scale/core/application_referrer_filter.rb]: @akostadinov the problem is what path traversal rack protection does with the path, see the [link](https://github.com/sinatra/sinatra/blob/v2.2.4/rack-protection/lib/rack/protection/path_traversal.rb#L28-L29) I posted above.

so, the ID `one/two\three` (in Ruby `"one/two\\three"`) is encoded by pisoni correctly into `one%2Ftwo%5Cthree`, but the path traversal converts it to `one/two/thee`, which is obviously wrong, 

```
/internal/services/500/applications/one%2Ftwo%5Cthree/keys/one%2Ftwo%5Cthree
=>
/internal/services/500/applications/one/two/thee/keys/one/two/thee
```

Comment (mayorova) [reply lib/3scale/core/application_referrer_filter.rb]: I think actually disabling path traversal is not going to cause a security issue, because we do not support static files, so I guess we can disable it.

However, I've been trying to disable this on the internal API, with no luck so far :upside_down_face:

Comment (jlledom) [reply lib/3scale/core/application_referrer_filter.rb]: Are `\` and `/` valid characters for app ids? IMO the right thing to do is to disallow those characters in app ids and keep path traversal enabled. I don't feel comfortable disabling security checks.

But if we already have app ids with slashes in the DB there's not much we can do. Disabling path traversal should be safe if we are sure Apisonator will never serve FS files, but how are you sure it won't? is there some config I can check?

Comment (jlledom) [reply lib/3scale/core/application_referrer_filter.rb]: > However, I've been trying to disable this on the internal API, with no luck so far 🙃

Is this test of any help?

https://github.com/sinatra/sinatra/blob/b626e2d82c23b4fde0b51782fd32ca27ccde1d1a/test/settings_test.rb#L614-L620

File: docker/docker-compose.apisonator.yml (hunk 3 above)
Comment (jlledom) [reply docker/docker-compose.apisonator.yml:15]: Shit, we should solve this in Apisonator, I don't think we should modify this file here. Better just set `APISONATOR_REL` to the last GA instead of `latest`.

Comment (jlledom) [reply docker/docker-compose.apisonator.yml:15]: Another thing that makes no sense IMO is having the `.env` file tracked by git. It's OK to track a `.env.example` but not the real one.

**Review by jlledom — CHANGES_REQUESTED** [review]

File: lib/3scale/core/application_referrer_filter.rb (hunk 1 above)
Comment (mayorova) [reply lib/3scale/core/application_referrer_filter.rb]: > Are `\` and `/` valid characters for app ids? IMO the right thing to do is to disallow those characters in app ids and keep path traversal enabled. I don't feel comfortable disabling security checks.

I had previously disallowed `/` in app IDs and left a note about it [here](https://github.com/3scale/porta/pull/3639). I had checked the production DB and confirmed that there were very few occurrences, and all - for old or test apps.

I didn't realize `\` would be problematic too, so it's currently allowed, but we can check the DB as well, and potentially disallow it.

> But if we already have app ids with slashes in the DB there's not much we can do. Disabling path traversal should be safe if we are sure Apisonator will never serve FS files, but how are you sure it won't? is there some config I can check?

I am not sure to be honest, I am trying to understand how sinatra is configured :grimacing:

File: docker/docker-compose.apisonator.yml (hunk 3 above)
Comment (mayorova) [reply docker/docker-compose.apisonator.yml:15]: @jlledom IMO, changing `RACK_ENV=test` is fine, because indeed we are using apisonator here just for test, not for development, so I don't see a problem.

Also, I think using `latest` is fine, because we probably want pisoni development to be aligned with the latest upstream, because SaaS goes faster than on-premises.

Re: `.env` file, in general I don't like the current setup at all. I think it's overly complicated with a complex (at least for me) `Makefile`, several docker-compose files, several Dockerfiles with different builder images. The base images are also outdated and we need to upgrade them.
So, I would revamp the whole thing and make it simpler and more maintainable. But... that's for another PR :wink:

Comment (mayorova) [reply docker/docker-compose.apisonator.yml:15]: I opened a JIRA: https://issues.redhat.com/browse/THREESCALE-10802
(description TBD)

File: lib/3scale/core/application_referrer_filter.rb (hunk 1 above)
Comment (jlledom) [reply lib/3scale/core/application_referrer_filter.rb]: > I had previously disallowed / in app IDs and left a note about it https://github.com/3scale/porta/pull/3639

Oh, I see you already faced the issue in that other PR

> I didn't realize \ would be problematic too, so it's currently allowed, but we can check the DB as well, and potentially disallow it.

This would be the ideal solution, if possible.

File: docker/docker-compose.apisonator.yml (hunk 3 above)
Comment (jlledom) [reply docker/docker-compose.apisonator.yml:15]: OK, fair enough

**Review by jlledom — APPROVED** [review]

File: lib/3scale/core/application_referrer_filter.rb (hunk 1 above)
Comment (mayorova) [reply lib/3scale/core/application_referrer_filter.rb]: In the meanwhile... https://github.com/3scale/apisonator/pull/366

**Review by jlledom — COMMENTED** [review]
It seems the last commit broke some tests

**Comment by mayorova** [comment]
> It seems the last commit broke some tests

Well, that's because this functionality depends on https://github.com/3scale/apisonator/pull/366 :sweat_smile: 

So if we decide not to merge it, I'll remove this commit.

File: lib/3scale/core/application_referrer_filter.rb (hunk 1 above)
Comment (akostadinov) [reply lib/3scale/core/application_referrer_filter.rb]: I'm ok to disable this. Seems to affect just this endpoint and we don't seem to serve files. Also relying on this protection to serve file would be very irresponsible.

**Review by akostadinov — APPROVED** [review]
